enterprise/server/scheduling/task_router/task_router.go

```diff
@@ -488,3 +502,58 @@ func getFirstOutput(cmd *repb.Command) string {
 	}
 	return ""
 }
+
+// persistentWorkerRouter generates routing keys based on:
+//   - groupID
+//   - remote_instance_name
+//   - persistentWorkerKey
+type persistentWorkerRouter struct{
+	env environment.Env
+}
+
+func (r *persistentWorkerRouter) Applies(ctx context.Context, params routingParams) bool {
+	if platform.FindValue(params.platform, "persistentWorkerKey") == "" {
+		return false
+	}
+	if exp := r.env.GetExperimentFlagProvider(); exp != nil {
+		return exp.Boolean(ctx, "remote_execution.enable_persistent_worker_routing", false)
+        }
+	return false
+}
+
+func (persistentWorkerRouter) preferredNodeLimit(_ routingParams) int {
+	// Intentionally set to the number of probes so that
+	// any task with a persistent worker key gets a hot node.
+	return 3
+}
+
+func (persistentWorkerRouter) routingKey(params routingParams) (string, error) {
+	parts := []string{"task_route", params.groupID}
+
+	if params.remoteInstanceName != "" {
+		parts = append(parts, params.remoteInstanceName)
+	}
+	parts = append(parts, platform.FindValue(params.platform, "persistentWorkerKey"))
+
+	return strings.Join(parts, "/"), nil
+}
+
+func (s persistentWorkerRouter) RoutingInfo(params routingParams) (int, []string, error) {
+	nodeLimit := s.preferredNodeLimit(params)
+	key, err := s.routingKey(params)
+	return nodeLimit, []string{key}, err
+}
+
+func (s persistentWorkerRouter) RankNodes(params routingParams, nodes []interfaces.ExecutionNode) []interfaces.RankedExecutionNode {
+	key, err := s.routingKey(params)
+	if err == nil {
+		src := rand.NewSource(int64(hash.MemHashString(key)))
```

Comment:
do we have to sort the nodes before shuffling so that it's deterministic? idk if the originally passed in nodes have a stable order